### PR TITLE
Epic47

### PR DESCRIPTION
**Pull request by shaliniib**

### CODE REVIEW

File: simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java [hunk 1]
```diff
@@ -626,9 +632,393 @@ public void setWidths(int[] widths) {
...
+
+	/** 
+	 * 
+	 * @param sheet
+	 * 			  sheet 			
+	 * @param columnName
+	 * 			  columnName
+	 * @return array of column values of primitive data type  
+	 */
+	public Object[] columnAsArray(String columnName) {
```
Comment (archetana) [inline simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java]: replace object with type <T> T[]

File: simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java [hunk 2]
```diff
@@ -626,9 +632,393 @@ public void setWidths(int[] widths) {
...
+		return list;
+	}
+
+	/**
+	 * 
+	 * @param sheet
+	 * @param columnName
+	 * @return set of column values
+	 */
+	public Set<Object> columnAsSet(String columnName) {
```
Comment (archetana) [inline simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java]: combine the columnAsSet and columnAsList to columnsAsCollection

File: simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java [hunk 3]
```diff
@@ -626,9 +632,393 @@ public void setWidths(int[] widths) {
...
+	}
+
+	/**
+	 * 
+	 * @param sheet
+	 * @param keyColumnName
+	 * @param valueColumnName
+	 * @return 2 column values as a map with 1 column data as keys and other column data as values
+	 */
+	public Map<String,Object> columnsAsMap(String keyColumnName, String valueColumnName) {
```
Comment (archetana) [inline simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java]: signature changes
int columnsAsMap(String keyColumnName, String valueColumnName,Map<K,V> map)

File: simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java [hunk 4]
```diff
@@ -626,9 +632,393 @@ public void setWidths(int[] widths) {
...
+	}
+
+	/**
+	 * 
+	 * @param sheet
+	 * @param className
+	 * 				fully qualified class name
+	 * @return List of entity objects 
+	 */
+	public List<Object> toList(String className) {
```
Comment (archetana) [inline simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java]: public int toList(List<T> listObject,Class?)

File: simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java [hunk 5]
```diff
@@ -626,9 +632,393 @@ public void setWidths(int[] widths) {
...
+					for (int j = 0; j < columnNames.length; j++) {
+						/* 
+						 * sets value to corresponding field of class instance 
+						 */
+						if (field.getName().equalsIgnoreCase(columnNames[j])) {							
+							/*
+							 * We expects the caller to make sure the field names of the class and data sheet column names to be same
+							 */
+							field.setAccessible(true);
+							SheetUtil.setAttribute(obj, field.getName(), row[j].toString(), false);
```
Comment (archetana) [inline simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java]: move setAttribute method to reflectUtil

File: simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java [hunk 6]
```diff
@@ -626,9 +632,393 @@ public void setWidths(int[] widths) {
...
+	}
+
+	/**
+	 * 
+	 * @param arr
+	 * @param columnName
+	 * @return MultiRowsSheet
+	 */
+	public static MultiRowsSheet toDatasheet(Object[] arr,String columnName) {
+		Class<?> cls = arr[0].getClass();
```
Comment (archetana) [inline simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java]: check arr length and null

File: simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java [hunk 7]
```diff
@@ -626,9 +632,393 @@ public void setWidths(int[] widths) {
...
+		for(Object key:map.keySet()){
+			Value[] row = new Value[2];
+			row[0] = Value.newTextValue(key.toString());
+			row[1] = Value.newTextValue(map.get(key).toString());
+			multirowsSheet.addRow(row);
+		}
+		return multirowsSheet;
+	}
+
+	private Object valueToPrimitive(Value value){
```
Comment (archetana) [inline simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java]: move this method to ValueType

File: simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java [hunk 8]
```diff
@@ -626,9 +632,393 @@ public void setWidths(int[] widths) {
...
+				return value.toString();
+			}
+			}
+		} catch (InvalidValueException e) {
+			throw new ApplicationError(e.getMessage());
+		}
+	}
+	
+	private static ValueType getType(Class<?> type) {
+		if (type.equals(String.class)) {
```
Comment (archetana) [inline simplity/kernel/src/main/java/org/simplity/kernel/data/MultiRowsSheet.java]: Use Value.parseObject(Object)

**Comment by shaliniib** [comment]
Modified as per review comments.

**Pull request closed without merging**